### ISSUE TITLE
weird output in dirlines PoC

### ISSUE THREAD

**Issue by joseluis** [dankamongmen/notcurses#1625]
I was searching in the PoC for some usage examples, and I only found it in the examples `dirgb` & `dirlines`, although commenting out every lines related to the default color doesn't make the output any different... So I'm a bit puzzled about that...

But the strange thing I found is that, independently of those lines (commented in/out & recompiled, etc.), the output started being like this, the first time I run it:
![image](https://user-images.githubusercontent.com/5832068/116926667-eab9a080-ac5a-11eb-9092-573d0327aca2.png)

the second one evolved to this:
![image](https://user-images.githubusercontent.com/5832068/116927548-1ab57380-ac5c-11eb-84b6-79a5ccdc0c0e.png)

and from the third time till now it always shows like this, in every terminal:
![image](https://user-images.githubusercontent.com/5832068/116927562-1db06400-ac5c-11eb-869a-06432b5c455c.png)

**Comment by dankamongmen** [comment]
i can reproduce this. it's probably just a bug in `dirlines`, not a general thing, but i'll look into it and fix it regardless.

**Comment by dankamongmen** [comment]
so `ncdirect_box()` never emits newlines, and thus we don't get any scrolling action here. i think at `ncdirect_cursor_down()` we'd need check whether we're on the last line of the terminal, and emit a newline if so. perhaps we can just always emit a newline there?

**Comment by dankamongmen** [comment]
maybe we can check our position before and after the down, and if it's the same, emit the newline?

**Comment by dankamongmen** [comment]
hrmmm. the obvious (if gross) approach of querying for cursor position ran into some problems, not getting a response at the end. very odd.

**Comment by dnkl** [comment]
If you're in "raw" mode, emitting `\n` is equivalent to moving the cursor down one step, but also scrolling the output if necessary. I.e. you shouldn't need checking the cursor position.

**Comment by dankamongmen** [comment]
> If you're in "raw" mode, emitting `\n` is equivalent to moving the cursor down one step, but also scrolling the output if necessary. I.e. you shouldn't need checking the cursor position.

i can't assume i'm in raw mode, though, nor unilaterally place us in it :(

**Comment by dankamongmen** [comment]
hrmmm, i suppose i could for the duration of writing

**Comment by dankamongmen** [comment]
and termios operations are reversible....hrmmm, yeah that could work, thanks a lot (as always) for the tip @dnkl!

**Comment by dankamongmen** [comment]
hrmmm perhaps i'm doing something wrong, but some quick experimentation is not demonstrating this behavior (which i also don't see documented in `termios(4)`):

```c
[schwarzgerat](0) $ cat ../src/poc/rawmode.c 
#define _DEFAULT_SOURCE
#include <termios.h>
#include <unistd.h>
#include <notcurses/direct.h>

int main(void){
  int r = EXIT_SUCCESS;
  struct ncdirect* nc = ncdirect_init(NULL, NULL, 0);
  if(nc == NULL){
    return EXIT_FAILURE;
  }
  struct termios term;
  cfmakeraw(&term);
  fprintf(stdout, ">>>>>>>>>>>>>\n");
  fprintf(stdout, ">>>>>>>>>>>>>\n");
  tcsetattr(fileno(stdout), TCSADRAIN, &term);
  ncdirect_stop(nc);
  return r;
}
[schwarzgerat](0) $ ./rawmode 
>>>>>>>>>>>>>
>>>>>>>>>>>>>
[schwarzgerat](0) $ 
```

going by your claim, i would expect the second line to start where the first line ends.

**Comment by dnkl** [comment]
`cfmakeraw()` doesn't set the termios attributes, it only updates the struct you pass to it. Try something like this:
```c
#define _DEFAULT_SOURCE
#include <termios.h>
#include <unistd.h>
#include <notcurses/direct.h>

int main(void){
  int r = EXIT_SUCCESS;
  struct ncdirect* nc = ncdirect_init(NULL, NULL, 0);
  if(nc == NULL){
    return EXIT_FAILURE;
  }
  struct termios term;
  tcgetattr(fileno(stdout), &term);
  struct termios raw = term;
  cfmakeraw(&raw);
  tcsetattr(fileno(stdout), TCSADRAIN, &raw);
  fprintf(stdout, ">>>>>>>>>>>>>\n");
  fprintf(stdout, ">>>>>>>>>>>>>\n");
  tcsetattr(fileno(stdout), TCSADRAIN, &term);
  ncdirect_stop(nc);
  return r;
}
```
```sh
✓ daniel@mini ~/src/foot/bld/debug% /tmp/test
>>>>>>>>>>>>>
             >>>>>>>>>>>>>
                          % 
```

**Comment by dankamongmen** [comment]
bah, i was misled by

```
       cfmakeraw() sets the terminal to something like the "raw" mode of the old  Ver‐
       sion  7  terminal driver: input is available character by character, echoing is
       disabled, and all special processing of terminal input and output characters is
       disabled.  The terminal attributes are set as follows....
```

but `strace` reveals the treachery. adding a `tcsetattr()` following the `cftermraw()` does indeed get the desired effect. Sweet!

**Comment by dnkl** [comment]
note the `tcgetattr()` before it too; `cfmakeraw()` doesn't built one up from the scratch, but modifies an existing one. Though in most cases I suppose a zero-initialized struct will do.

**Comment by dankamongmen** [comment]
so @dnkl i am in that case tempted to replace the content of `ncdirect_cursor_down()` (which currently uses the `cud` terminfo capability) entirely with a

```c
tcgetattr(fileno(stdout), &term);
cfmakeraw(&raw);
tcsetattr(nc->ttyfd, TCSADRAIN, &raw);
while(n--){
  fputc_unlocked('\n', nc->ttyfp);
}
tcsetattr(nc->ttyfd, TCSADRAIN, &term);
```

plus error checks. i don't love the idea of three system calls for every invocation (and even that might require disabling line buffering, lest each of those `fputc()`s take a flush), but that'll make it work.

**Comment by dankamongmen** [comment]
> note the `tcgetattr()` before it too; `cfmakeraw()` doesn't build one up from the scratch, it modifies an existing one. Though in most cases I suppose a zero-initialized struct will do.

ja

**Comment by dankamongmen** [comment]
hrmmm does `cud` work this way as well if one is in raw mode?

**Comment by dankamongmen** [comment]
if so, that's a nice O(1) solution; i shall check (doubting it though, as the CR is part of the line discipline, right?)

**Comment by dnkl** [comment]
A couple of other solutions you might want to investigate; not sure how well supported these are:

* `\v`
* `\f`
* `\ED`

Sorry I don't have their fancy names, I just did a quick grep in the foot source code. All of these are implemented exactly like `\n` in raw mode (in foot at least).

**Comment by dankamongmen** [comment]
> A couple of other solutions you might want to investigate; not sure how well supported these are:
> 
> * `\v`
> * `\f`
> * `\ED`
> 
> Sorry I don't have their fancy names, I just did a quick grep in the foot source code. All of these are implemented exactly like `\n` in raw mode (in foot at least).

`\f` is full form feed, right? i.e. a page at a time

**Comment by dnkl** [comment]
That's its name, yes. But quick testing in XTerm shows it moves the cursor down one step... not a full page.

**Comment by dnkl** [comment]
> hrmmm does cud work this way as well if one is in raw mode?

No, don't think so; this one only moves the cursor, it never scrolls.

**Comment by dankamongmen** [comment]
> > hrmmm does cud work this way as well if one is in raw mode?
> 
> No, don't think so; this one only moves the cursor, it never scrolls.

indeed, no love (as expected)

i could keep an array around of repeated `\n` and use `printf("%*s", num)` but that's a bit silly. zee loop it is.

**Comment by dnkl** [comment]
> All of these are implemented exactly like \n in raw mode (in foot at least).

Let me rephrase that; all of them behave just like `\n`. Raw mode doesn't really have anything to do with it, it's just that `\n` is translated to `\n\r` by the TTY driver in cooked mode.

What I'm thinking is that these alternatives doesn't need you to switch between raw and cooked mode.

**Comment by dankamongmen** [comment]
works perfectly! bravo, @dnkl! you're the fuckin' man! no idea where i would have picked that up otherwise...thanks a lot for watching and dropping knowledge!

![2021-05-05-044601_801x1417_scrot](https://user-images.githubusercontent.com/143473/117116885-e4abe880-ad5c-11eb-9a08-d0acad3bbe59.png)

**Comment by dnkl** [comment]
> What I'm thinking is that these alternatives don't need you to switch between raw and cooked mode.

Let me push this one once more, in case it got lost in the email notification storm ;)

> thanks a lot for watching and dropping knowledge!

Happy to help :) A terminal isn't much without the applications running in it...

**Comment by dankamongmen** [comment]
> > What I'm thinking is that these alternatives don't need you to switch between raw and cooked mode.
> Let me push this one once more, in case it got lost in the email notification storm ;)

a big score with `\v`!

**Comment by dankamongmen** [comment]
Skål!

**Comment by dnkl** [comment]
> Skål

🍸
